#!/usr/bin/env Rscript

if(!require(optparse)){
        install.packages("optparse")
    library(optparse)
}
if(!require(data.table)){
        install.packages("data.table")
    library(data.table)
}
if(!require(svMisc)){
        install.packages("svMisc")
    library(svMisc)
}

option_list = list(
    
    make_option(c("-a", "--bam1"), 
                type="character", 
                default=NA,
                help="name of bam file: experiment (without .bam extension)", 
                metavar="character"),
    
    make_option(c("-b", "--bam2"), 
                type="character", 
                default=NA,
                help="name of bam file: control (without .bam extension)", 
                metavar="character"),
    
    make_option(c("-m", "--motifFile"), 
                type="character", 
                default=NA,
                help="File having location of motif; generated by Homer", 
                metavar="character"),
    
    make_option(c("-x", "--centreOfmotif"), 
                type="integer", 
                help="Desired centre of motif on positive DNA strand", 
                metavar="numeric"),
    
    make_option(c("-n", "--thread"), 
                type="integer", 
                default=2,
                help="Number of thread",metavar="numeric"),
    
    make_option(c("-p", "--prefix"), 
                type="character", 
                help="prefix of output files",
                metavar="character"),
    
    make_option(c("-w", "--pwm"), 
                type="character", 
                help="file containts position weight matrix of the motif",
                metavar="character"),
    
    make_option(c("-e", "--exprReads"), 
                type="integer",
                help="reads in the experiment"),
    
    make_option(c("-f", "--ctrlReads"), 
                type="integer", 
                help="reads in the control"),
    
    make_option(c("-s", "--spikeRatio"),
                type="double", 
                help="ratio of the spikeIn in experiment compared to control adjusted by their read numbers")
);

parser<- OptionParser(option_list=option_list)
opt <- parse_args(parser)

    bam1=opt$bam1
    bam2=opt$bam2
    mymotif=opt$motifFile
    mythread=opt$thread
    centre=opt$centreOfmotif
    prefix=opt$prefix
    pwm=opt$pwm
    exprReads=opt$exprReads
    ctrlReads=opt$ctrlReads
    spikeRatio=opt$spikeRatio

#---- data preparation: part2
d <- fread(paste(bam1,'.in',sep=''))
d <- data.frame(d)
    pos_d <- d[d[,6]=="+",]
    pos_d$end5 = pos_d[,8] - (pos_d[,2]+(centre-1))
    pos_d$end3 = pos_d[,9] - (pos_d[,2]+(centre-1))
    neg_d <- d[d[,6]=="-",]
    neg_d$end5 = neg_d[,9] - (neg_d[,2]+(centre-1))
    neg_d$end3 = neg_d[,8] - (neg_d[,2]+(centre-1))
    rbind(pos_d,neg_d) -> values
    values[order(values[,1],values[,2]),] -> values
    experiment = values
#--
d <- fread(paste(bam2,'.in',sep=''))
d <- data.frame(d)
    pos_d <- d[d[,6]=="+",]
    pos_d$end5 = pos_d[,8] - (pos_d[,2]+(centre-1))
    pos_d$end3 = pos_d[,9] - (pos_d[,2]+(centre-1))
    neg_d <- d[d[,6]=="-",]
    neg_d$end5 = neg_d[,9] - (neg_d[,2]+(centre-1))
    neg_d$end3 = neg_d[,8] - (neg_d[,2]+(centre-1))
    rbind(pos_d,neg_d) -> values
    values[order(values[,1],values[,2]),] -> values
    control = values
#--
d <- fread(mymotif)
d <- data.frame(d)
nrow(d) -> motifs_number
#--
experiment_reads=exprReads
control_reads=ctrlReads
spike_ratio=spikeRatio

#--- Nearest cut
print ('=== Finding nearest cut for each reads ')
print ('=== Experiment ')
    as.numeric(
        ifelse (
            (experiment$end5^2)^(1/2) < (experiment$end3^2)^(1/2) , 
            experiment$end5, 
            experiment$end3)
    ) -> experiment$V15
print ('=== Control ')
    as.numeric(
        ifelse (
            (control$end5^2)^(1/2) < (control$end3^2)^(1/2) , 
            control$end5, 
            control$end3)
    ) -> control$V15
    
    experiment$region=paste(experiment[,1],':',experiment[,2],'-',experiment[,3],sep='');
    control$region=paste(control[,1],':',control[,2],'-',control[,3],sep='');

#-- counting
print ('=== Counting cuts per region ')

    experiment[experiment$V15 > -101 & experiment$V15 < 101, ] -> experiment
    table(experiment$region,experiment$V15) -> mytable_expr
    mytable_expr[,as.character(seq(-100,100,1))]->mytable_expr
    
    #--
    control[control$V15 > -101 & control$V15 < 101, ] -> control
    table(control$region,control$V15) -> mytable_ctrl
    mytable_ctrl[,as.character(seq(-100,100,1))]->mytable_ctrl


print ('=== output of counts in table ')

    write.table(mytable_ctrl[rowSums(mytable_ctrl) > 1,],
                paste(prefix,'.counts.txt',sep=''),
                quote=F,
                sep='\t',
                col.names=F)
    write.table(mytable_expr[rowSums(mytable_expr) > 1,],
                paste(prefix,'.counts.txt',sep=''),
                quote=F,
                sep='\t',
                col.names=F)
#---- normalization

print ('=== Normalization ')
    common = intersect(rownames(mytable_ctrl),rownames(mytable_expr))
    specific2expr = setdiff(rownames(mytable_expr),rownames(mytable_ctrl))
    mytable_expr=1000000*((mytable_expr/experiment_reads)/motifs_number)
    mytable_ctrl = mytable_ctrl*spike_ratio
    mytable_ctrl=1000000*((mytable_ctrl/control_reads)/motifs_number)

print ('=== Comparing with control ')
    out = mytable_expr[common,] - mytable_ctrl[common,]
    out = rbind(out,mytable_expr[specific2expr,])
    write.table(out[rowSums(out) > 0,],
                paste(prefix,'.normalize.txt',sep=''),
                quote=F,
                sep='\t',
                col.names=F)
    z<-colSums(out)
    out_frequency=data.frame(Distance=as.numeric(names(z)),Frequency=as.numeric(z))
    write.table(out_frequency,
                paste(prefix,'.cutfrequency.txt',sep=''),
                quote=F,
                sep='\t',
                col.names=F)
    print(z)
    maximum=max(z)
    minimum=min(z)
    minimum=minimum+(minimum*0.05)

jpeg(paste(prefix,'.jpeg',sep=''),
     unit='in',
     res=300,
     height=4.5,
     width=4.5)

    plot(as.numeric(names(z)),
         z,
         ylim=c(minimum,maximum),
         type='n',
         xlab='Position (bps)',
         ylab='Cut frequency')
    
    cc = nrow ( read.table(pwm, skip = 1, 
                           header = F) )
    #cc = nchar(experiment[1,6])
    right_cc = cc - centre
    left_cc = -1 * (cc - (right_cc + 1))
    
    rect(left_cc,
         minimum-500,
         right_cc,
         maximum+500,
         col='grey80',
         border = NA)
    
    for(i in seq(-100,100,10)) {
        abline(v=i,lty=2,col='grey72')
    }
    
    lines(as.numeric(names(z)),
          z,
          cex=0.8,
          col='black',
          type='o')

dev.off()

print ('=== Finished ')
print ('=============================')

